### ISSUE TITLE
Link modules instead of cloning modules in the codegen in LLVM backends

### ISSUE THREAD

**Issue by lin-hitonami** [taichi-dev/taichi#5511]
**Concisely describe the proposed feature**
As described in #5282, the compilation of Taichi kernel becomes slow when there are many SNodeTrees. After profiling, we found that the process of cloning the modules take up most of the time. The cloning happen in three places. 
Two of them happen in the construction process of the struct module as described in #3146. 
https://github.com/taichi-dev/taichi/blob/77398de87e635e4de583de808137e8efe484e8a9/taichi/llvm/llvm_program.cpp#L138-L145
https://github.com/taichi-dev/taichi/blob/77398de87e635e4de583de808137e8efe484e8a9/taichi/llvm/llvm_context.cpp#L455
The other one happens when the LLVM codegen gets the struct module as the initial module of kernel module.
https://github.com/taichi-dev/taichi/blob/d00348c1acccda88c1823e60e151a78b1c4f1b8a/taichi/codegen/llvm/codegen_llvm.cpp#L318-L322

**Describe the solution you'd like (if any)**
We can store the runtime functions, the struct functions of the SNodeTrees, and the kernel functions in different modules, and link the needed modules at the end.

**Comment by lin-hitonami** [comment]
Using an altered version of the program in https://github.com/taichi-dev/taichi/issues/5282:
```
import taichi as ti
from tqdm import tqdm

if __name__ == "__main__":
    ti.init(ti.cpu)
    iterations = 2
    snode_tree_num = 511 # Or 512 without relaxing `constexpr int kMaxNumSnodeTreesLlvm` in taichi/inc/constants.h
    for i in range(iterations):
        snode_trees = []
        for j in tqdm(range(snode_tree_num)):
            x = ti.field(ti.f32)
            fb = ti.FieldsBuilder()
            fb.place(x)
            snodetree = fb.finalize()
            a = x[None]
            snode_trees.append(snodetree)
        for snodetree in snode_trees:
            snodetree.destroy()
```

the module splitting achieved ~7x speed up when the number of SNodeTree is 511.

The current master slows down from ~50iters/s to ~10iters/s, and https://github.com/taichi-dev/taichi/pull/5587 runs stably on ~70iters/s.

The output of master:
```
[Taichi] version 1.0.4, llvm 10.0.0, commit dc93ed8f, linux, python 3.10.4
[Taichi] Starting on arch=x64
100%|█████████████████████████████████████████| 511/511 [00:28<00:00, 18.01it/s]
100%|█████████████████████████████████████████| 511/511 [00:19<00:00, 26.49it/s]
```
The output of https://github.com/taichi-dev/taichi/pull/5587:
```
[Taichi] version 1.0.5, llvm 10.0.0, commit af9b94d7, linux, python 3.10.4
[Taichi] Starting on arch=x64
100%|█████████████████████████████████████████| 511/511 [00:07<00:00, 71.35it/s]
100%|████████████████████████████████████████| 511/511 [00:01<00:00, 257.81it/s]
```

**Comment by lin-hitonami** [comment]
> After getting this resolved, do you think there's still a need to limit the number of `SNodeTree`s by using `kMaxNumSnodeTreesLlvm`? I filed a feature request #4033 long time ago, and it seems to me now it's the right time to resolve it altogether.

This issue and #4033 are irrelevant. But #4033 can be the next step, let me look into it.

**Comment by ifsheldon** [comment]
I guess this issue can be closed and then #5282 can be closed as well? I am really looking forward to the new release!

**Comment by lin-hitonami** [comment]
Yes!
